cmd/kubeadm/app/phases/patchnode/patchnode.go

```diff
@@ -26,9 +26,11 @@ import (
 
 // AnnotateCRISocket annotates the node with the given crisocket
 func AnnotateCRISocket(client clientset.Interface, nodeName string, criSocket string) error {
-
-	klog.V(1).Infof("[patchnode] Uploading the CRI Socket information %q to the Node API object %q as an annotation\n", criSocket, nodeName)
-
+	klog.V(1).Infof("[patchnode] Creating the Node API object %q if missing", nodeName)
+	if err := apiclient.CreateNode(client, nodeName); err != nil {
```

Comment:
I find a little weird that `CreateNode` is called from two different places in the call tree (even if it succeeds if the node already exists).

Reply:
I have to dig into the PR in a proper way, but I also fail to see how `CreateNode` will create the node object. Shouldn't the node object have already been created when we reach this point?

Reply:
no, the TLS bootstrap delays the Node creation.
the manual pre-creation saves a few dozen seconds, we can wait, but this way it's faster and i don't see any issues with it.

the CreateNode is called in two different locations because i didn't want to add it as part of the node patching methods.

Reply:
Thank you for the explanation @neolit123, just to double check that I understand:

> the manual pre-creation saves a few dozen seconds (or maybe just a dozen), we can wait, but this way it's faster and i don't see any issues with it.

If we removed the `CreateNode` calls, everything should still be working as previously, right? (We are faster now because of the creation of the node "placeholder", but nothing would break if we didn't pre-create it)

Reply:
no, patch node calls will fail, they try patching for a while, but the node object pre-create ensures that they succeed.

you can try downloading this patch and try removing the CreateNode calls.

Reply:
on the other hand, if you remove the CRI socket and markcontrolplane patch calls init should succeed.

Reply:
What I don't understand I guess is how this PR changes the semantics on how the kubelet registers. I assume this PR does not change anything like that (the kubelet will try to register itself as always with https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/kubelet_node_status.go#L53-L80. Or does this change with this PR in a way I'm missing?

I'm fine pre-creating the node object, the only issue I foresee with this strategy is if the TLS bootstrap fails, we will have a dangling node object unless we rollback its creation with kubeadm if the process failed. The dangling object in kubeadm is not an issue because on retry it's idempotent, but in the past we had a PR open to fail if a node object existed with the same name (and some third parties can have this strategy as well).

Reply:
i think if the TLS boostrap fails the init / join process will exit before our placeholder Node object is created.